[RISCV][GISel] Add support for G_IS_FPCLASS in F and D extensions

mshockwave opened the pull request:
Add legalizer, regbankselect, and isel supports for floating point version of G_IS_FPCLASS.

mshockwave committed:
[RISCG][GISel] Add support for G_IS_FPCLASS in F and D extensions

Add legalizer, regbankselect, and isel supports for floating point
version of G_IS_FPCLASS.

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -270,6 +279,17 @@ bool RISCVLegalizerInfo::legalizeCustom(LegalizerHelper &Helper,
            LegalizerHelper::Legalized;
   }
+  case TargetOpcode::G_IS_FPCLASS: {
+    // Turn LLVM IR's floating point classes to that in RISCV,
```
If we change the immediate here, would that confuse generic combines that look at G_IS_FPCLASS?

mshockwave replied:
Good point. I think right now none of the combiner rules, generic or not, is targeting G_IS_FPCLASS. And I feel like future post-legalizer combiners have to take this in mind.

topperc replied:
I think we should move the translation to RISC-V encoding to instruction selection.

mshockwave replied:
> I think we should move the translation to RISC-V encoding to instruction selection.

Will do!

mshockwave replied:
Done

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,30 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+                 .addImm(MaskImm);
+  if (!And.constrainAllUses(TII, TRI, RBI))
+    return false;
+
```
G_IS_FPCLASS is supposed to produce 0/1 in the LSB and have all other bits as 0. This code will can produce values between 0-1023.

mshockwave replied:
I think G_IS_FPCLASS's [doc](https://llvm.org/docs/GlobalISel/GenericOpcode.html#g-is-fpclass) states that any non-zero values is allowed to represent true: "...Returns non-zero (true) or zero (false). It’s target specific whether a true value is 1, ~0U, or some other non-zero value."

topperc replied:
This test case miscompiles

```
define zeroext i1 @foo(float %x) {                                               
  %a = call i1 @llvm.is.fpclass.f32(float %x, i32 1)                             
  ret i1 %a                                                                      
}                                                                                
                                                                                 
declare i1 @llvm.is.fpclass.f32(float, i32)  
```

The resulting code is
```
        fclass.s        a0, fa0
        andi    a0, a0, 256
        andi    a0, a0, 1
        ret
```

The second andi clears the bit from the first andi.

topperc replied:
I believe that the result should adhere to `TLI->getBooleanContents` which `ZeroOrOneBooleanContent`. That's the target specific part.

mshockwave replied:
You're right, I was using ANYEXT rather than ZEXT in my legalizer test and missed the ANDI it generated. 
It's fixed now.

mshockwave replied:
> I believe that the result should adhere to `TLI->getBooleanContents` which returns `ZeroOrOneBooleanContent` for RISC-V. That's the target specific part.

I didn't see this comment before pushing the latest update. Will fix in the next update

topperc replied:
I don't expect you to use `getBooleanContents`. It will always be ZeroOrOneBooleanContent for RISC-V. I'm fine with assuming that. If you were writing target independent code you would need to use it.

tschuett commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,30 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+bool RISCVInstructionSelector::selectIsFPClass(MachineInstr &MI,
+                                               MachineIRBuilder &MIB,
+                                               MachineRegisterInfo &MRI) const {
+  Register CheckResult = MI.getOperand(0).getReg();
```
Would a `GISFPLASS` simplify the code?

mshockwave replied:
Done

mshockwave committed:
Move transformation of FPClass mask from legalizer to isel

And addressed some comments.

mshockwave committed:
fixup! Move transformation of FPClass mask from legalizer to isel

mshockwave committed:
Make sure the result only set/clear the LSB

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,56 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+  unsigned NewOpc = MRI.getType(Src).getSizeInBits() == 32 ? RISCV::FCLASS_S
+                                                           : RISCV::FCLASS_D;
+
+  // Turn LLVM IR's floating point classes to that in RISCV,
```
RISCV -> RISC-V.

mshockwave replied:
Done

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,56 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+  if (!FClass.constrainAllUses(TII, TRI, RBI))
+    return false;
+
+  // If there is only a single bit set in the mask, lower to SLLI + SRLI.
```
There's a second special case. You can use ANDI by itself if FClassMask is 1.

mshockwave replied:
Done

mshockwave committed:
Address review comments

And add another special case where FCLASS mask is one.

mshockwave commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,56 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+  // If there is only a single bit set in the mask, lower to SLLI + SRLI.
+  if (FClassMask.popcount() == 1) {
+    Register SLResult = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    auto SLLI = MIB.buildInstr(RISCV::SLLI, {SLResult}, {FClassResult})
```
Done

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:
```diff
@@ -948,6 +952,61 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+
+  if (FClassMask == 1) {
+    // We don't need to generate additional instructions if Mask is 1, as the
+    // `ANDI rs, 1` that follows will suffice.
```
I think we have to make an ANDI. It's not guaranteed to exist. If computeKnownBits ran on the G_ISFPCLASS, it would know that G_ISFPCLASS produces 0/1 and delete any existing ANDI.

arsenm replied:
I don't exactly follow why this needs post processing, but my feeling is if you're introducing regular bit instructions after this result you'd be better off doing this in custom legalization

mshockwave replied:
> I think we have to make an ANDI. It's not guaranteed to exist. If computeKnownBits ran on the G_ISFPCLASS, it would know that G_ISFPCLASS produces 0/1 and delete any existing ANDI.

Done

mshockwave replied:
> I don't exactly follow why this needs post processing, but my feeling is if you're introducing regular bit instructions after this result you'd be better off doing this in custom legalization

I think some contexts have been shifted around, which lines / post processing are you referring here?

topperc replied:
I think we should have a RISCV::G_FCLASS node and use custom legalization to turn G_FPCLASS into that plus G_AND and G_ICMP during legalization. Then we can use the existing isel patterns for RISCVISD::FPCLASS(which I think should be named RISCVISD::FCLASS so we should fix that too) by marking G_FCLASS as equivalent.

mshockwave committed:
fixup! Address review comments

mshockwave committed:
Create a generic instruction for G_IS_FPCLASS

Creating a generic instruction, RISCV::G_FCLASS, so that we can reuse
the SelectionDAG patterns during ISel.

mshockwave committed:
fixup! Create a generic instruction for G_IS_FPCLASS

mshockwave commented:
Now the legalizer generates a `RISCV::G_FCLASS` instruction that will reuse the codegen patterns from `RISCVISD::FPCLASS`. 

Note that RISCV's GISel currently lacks patterns to optimize the cases shown in the `is_fpclass_f*_onehot` and `is_fpclass_f*_one` functions. Namely, cases where FCLASS mask only has a single bit and FCLASS mask == 1.
But I still keep those cases in the test files so that it will be more obvious when those optimizations are landed in the future.

mshockwave committed:
fixup! Create a generic instruction for G_IS_FPCLASS

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -223,6 +224,8 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
       })
       .clampScalar(0, sXLen, sXLen);
 
+  getActionDefinitionsBuilder(G_IS_FPCLASS).custom();
```
Please rebase and then use `customIf(typeIsScalarFPArith(1, ST)` so that we don't do the custom lowering if F or D isn't enabled.

mshockwave replied:
Rebased and fixed.

topperc commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -21,9 +22,8 @@
-  const unsigned XLen = ST.getXLen();
-  const LLT sXLen = LLT::scalar(XLen);
+RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
+    : STI(ST), XLen(STI.getXLen()), sXLen(LLT::scalar(XLen)) {
```
Where are XLen and sXLen used?

mshockwave replied:
I forgot to update after factoring out XLen and sXLen, it's fixed now.

mshockwave committed:
Rebase and address comments in the legalizer

mshockwave commented on llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:
```diff
@@ -215,6 +215,7 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
                 typeIs(1, s32)(Query));
       });
 
+  // TODO: Support vector version of G_IS_FPCLASS.
```
Fixed

topperc commented on llvm/lib/Target/RISCV/RISCVInstrGISel.td:
```diff
@@ -0,0 +1,26 @@
+  let Namespace = "RISCV";
+}
+
+// Pseudo equivalent to a RISCVISD::FPCLASS.
```
Can you rename RISCVISD::FPCLASS to RISCVISD::FCLASS as pre or post commit?

mshockwave replied:
Renamed in https://github.com/llvm/llvm-project/commit/e096732307d1eb9476fc2b88dfa12586250c8111.  I've also updated this patch

topperc reviewed (approved):
LGTM

mshockwave committed:
Rename RISCVISD::FPCLASS
